// @atproto/syntax/dist/aturi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AtUri = exports.ATP_URI_REGEX = void 0;
const tslib_1 = require("tslib");
const at_identifier_js_1 = require("./at-identifier.js");
const did_js_1 = require("./did.js");
const nsid_js_1 = require("./nsid.js");
const recordkey_js_1 = require("./recordkey.js");
tslib_1.__exportStar(require("./aturi_validation.js"), exports);
exports.ATP_URI_REGEX = 
// proto-    --did--------------   --name----------------   --path----   --query--   --hash--
/^(at:\/\/)?((?:did:[a-z0-9:%-]+)|(?:[a-z0-9][a-z0-9.:-]*))(\/[^?#\s]*)?(\?[^#\s]+)?(#[^\s]+)?$/i;
//                       --path-----   --query--  --hash--
const RELATIVE_REGEX = /^(\/[^?#\s]*)?(\?[^#\s]+)?(#[^\s]+)?$/i;
class AtUri {
    hash;
    host;
    pathname;
    searchParams;
    constructor(uri, base) {
        const parsed = base !== undefined
            ? typeof base === 'string'
                ? Object.assign(parse(base), parseRelative(uri))
                : Object.assign({ host: base.host }, parseRelative(uri))
            : parse(uri);
        (0, at_identifier_js_1.ensureValidAtIdentifier)(parsed.host);
        this.hash = parsed.hash ?? '';
        this.host = parsed.host;
        this.pathname = parsed.pathname ?? '';
        this.searchParams = parsed.searchParams;
    }
    static make(handleOrDid, collection, rkey) {
        let str = handleOrDid;
        if (collection)
            str += '/' + collection;
        if (rkey)
            str += '/' + rkey;
        return new AtUri(str);
    }
    get protocol() {
        return 'at:';
    }
    get origin() {
        return `at://${this.host}`;
    }
    get did() {
        const { hostname } = this;
        (0, did_js_1.ensureValidDid)(hostname);
        return hostname;
    }
    get hostname() {
        return this.host;
    }
    set hostname(v) {
        (0, at_identifier_js_1.ensureValidAtIdentifier)(v);
        this.host = v;
    }
    get search() {
        return this.searchParams.toString();
    }
    set search(v) {
        this.searchParams = new URLSearchParams(v);
    }
    get collection() {
        return this.pathname.split('/').filter(Boolean)[0] || '';
    }
    get collectionSafe() {
        const { collection } = this;
        (0, nsid_js_1.ensureValidNsid)(collection);
        return collection;
    }
    set collection(v) {
        (0, nsid_js_1.ensureValidNsid)(v);
        this.unsafelySetCollection(v);
    }
    unsafelySetCollection(v) {
        const parts = this.pathname.split('/').filter(Boolean);
        parts[0] = v;
        this.pathname = parts.join('/');
    }
    get rkey() {
        return this.pathname.split('/').filter(Boolean)[1] || '';
    }
    get rkeySafe() {
        const { rkey } = this;
        (0, recordkey_js_1.ensureValidRecordKey)(rkey);
        return rkey;
    }
    set rkey(v) {
        (0, recordkey_js_1.ensureValidRecordKey)(v);
        this.unsafelySetRkey(v);
    }
    unsafelySetRkey(v) {
        const parts = this.pathname.split('/').filter(Boolean);
        parts[0] ||= 'undefined';
        parts[1] = v;
        this.pathname = parts.join('/');
    }
    get href() {
        return this.toString();
    }
    toString() {
        let path = this.pathname || '/';
        if (!path.startsWith('/')) {
            path = `/${path}`;
        }
        let qs = '';
        if (this.searchParams.size) {
            qs = `?${this.searchParams.toString()}`;
        }
        let hash = this.hash;
        if (hash && !hash.startsWith('#')) {
            hash = `#${hash}`;
        }
        return `at://${this.host}${path}${qs}${hash}`;
    }
}
exports.AtUri = AtUri;
function parse(str) {
    const match = str.match(exports.ATP_URI_REGEX);
    if (!match) {
        throw new Error(`Invalid AT uri: ${str}`);
    }
    return {
        host: match[2],
        hash: match[5],
        pathname: match[3],
        searchParams: new URLSearchParams(match[4]),
    };
}
function parseRelative(str) {
    const match = str.match(RELATIVE_REGEX);
    if (!match) {
        throw new Error(`Invalid path: ${str}`);
    }
    return {
        hash: match[3],
        pathname: match[1],
        searchParams: new URLSearchParams(match[2]),
    };
}
